arraycontext/impl/pytato/batched_einsum/utils.py

```diff
@@ -0,0 +1,647 @@
+        raise NotImplementedError(
+            "'apply_kennedy_fusion_with_batched_einsum_extension'"
+            " does not support multiple callee kernels (yet).")
+
+    kernel = t_unit.default_entrypoint
+
+    assert all(len(kernel.iname_to_insns()[iname]) <= 1
+               for iname in kernel.all_inames())
+
+    # A bucket by a tagged einsum and it's position in the einsum.
+    bucket_to_inames: Dict[Tuple[EinsumWithAxesTagged, fnsm.EinsumAxisAccess],
+                           Set[str]] = {}
+
+    for insn in kernel.instructions:
+        if isinstance(insn, lp.Assignment):
+            # {{{ get matching einsum/subst_map
+
+            if insn.reduction_inames():
+                einsum, _ = fnsm.get_a_matched_einsum(
+                    t_unit, insn_match=lp_match.Id(insn.id))
+                einsum = fnsm.canonicalize_einsum(einsum)
+                subst_map = fnsm.match_t_unit_to_einsum(
+                    t_unit, einsum, insn_match=lp_match.Id(insn.id))
+            else:
+                # we treat any non-reduction einsum as a copy-einsum
+                assignee = insn.assignee
+                if isinstance(assignee, prim.Variable):
+                    lpy_dim_names = []
+                else:
+                    assert isinstance(assignee, prim.Subscript)
+                    lpy_dim_names = [index.name for index in assignee.index_tuple]
+
+                dim_lengths = [kernel.get_constant_iname_length(dim_name)
+                               for dim_name in lpy_dim_names]
+                if len(lpy_dim_names) > 26:
+                    raise ValueError("Batched Einsum Actx supports upto"
+                                     "26-dimensions")
+                einsum_dim_names = [chr(97+idim)
+                                    for idim in range(len(lpy_dim_names))]
+                einsum = fnsm.einsum(
+                    f"{''.join(einsum_dim_names)}->{''.join(einsum_dim_names)}",
+                    # purposefully fix dtype=F64, since for such expression we are
+                    # imprecise on purpose.
+                    fnsm.array(shape=dim_lengths, dtype="float64"),
+                )
+                einsum = fnsm.canonicalize_einsum(einsum)
+                subst_map = {
+                    einsum.index_names[fnsm.FreeAxis(idim)]: lpy_dim_names[idim]
+                    for idim in range(len(einsum_dim_names))}
+            # }}}
+
+            idx_tags: Dict[fnsm.EinsumAxisAccess, Tag] = {}
+            for acc_descr, name_in_einsum in einsum.index_names.items():
+                lpy_iname = subst_map[name_in_einsum]
+                lpy_iname_tag, = kernel.iname_tags_of_type(lpy_iname, tag_t)
+                idx_tags[acc_descr] = lpy_iname_tag
+
+            tagged_einsum = EinsumWithAxesTagged(einsum, Map(idx_tags))
+
+            for acc_descr, name_in_einsum in einsum.index_names.items():
+                bucket = (tagged_einsum, acc_descr)
```

Comment:
I am trying to understand the logic here. This appears to be behind issues that @majosm is encountering when trialing this array context in mirgecom, it appears to want to fuse loops of different lengths. And if there are multiple matches to an einsum template within a Loopy kernel, it seems that this will happily toss all inames in analogous "access roles" into the same bucket, irrespective of whether they have the same length.

Beyond the obvious correctness concern, I also wonder about the thought process behind "toss everybody into one fusible bucket"... there might be many many matches within a big compile. All those loops would get fused, resulting in a kernel that may access many different variables, resulting in large amounts of state space use. I didn't see logic to pry these loops back apart if the batch-of-einsums gets too big. What's the intention here? I would be grateful if you could shed some light. Thanks!

Reply:
> it appears to want to fuse loops of different lengths

It cannot fuse loops with different lengths since the `tagged_einsum` would be different for those inames.

> And if there are multiple matches to an einsum template within a Loopy kernel,

We map an iname to a `(normalized_einsum, acc_descr_of_the_iname_in_the_normalized_einsum)`, this tuple should be unique to an einsum. 


> I didn't see logic to pry these loops back apart if the batch-of-einsums gets too big.

We would rely on a tuned `feinsum` transformation to determine of the suitable loop-distribution factor.

Reply:
> It cannot fuse loops with different lengths since the `tagged_einsum` would be different for those inames.

Could it be that the "template einsum"  is using a parametric size, and that's not being specialized for matching purposes? It's definitely trying to merge loops of different length in a particular situation for which @majosm unearthed a reproducer (https://github.com/illinois-ceesd/mirgecom/issues/967). I'll do some more digging.

> We would rely on a tuned feinsum transformation to determine of the suitable loop-distribution factor.

That's a sensible approach.

Reply:
> Could it be that the "template einsum" is using a parametric size, and that's not being specialized for matching purposes?

Yep. That's exactly what happened over here. (Thanks!). Loop Fusion should also have been less dumb, see: https://gist.github.com/kaushikcfd/c2707106c47f1fd19d70dfcfd77be845. I'll fix both these things.

Reply:
See the failing reproducer in [e4cf24b](https://github.com/inducer/arraycontext/pull/217/commits/e4cf24b8c1f64f1734cc223d55531d59024bf515) and the associated fix in caaa471.

Thanks for the report!

Reply:
Didn't mean to make you do work! But nonetheless thanks for looking into it, and for coming up with a fix. :slightly_smiling_face: